Objects/typeobject.c

```diff
@@ -6166,23 +6286,45 @@ type_setattro(PyObject *self, PyObject *name, PyObject *value)
 
     PyObject *dict = type->tp_dict;
     if (dict == NULL) {
-        // We don't just do PyType_Ready because we could already be readying
-        BEGIN_TYPE_LOCK();
-        dict = type->tp_dict;
-        if (dict == NULL) {
+        // This is an unlikely case.  PyType_Ready has not yet been done and
+        // we need to initialize tp_dict.  We don't just do PyType_Ready
+        // because we could already be readying.
+        types_stop_world();
+        if (type->tp_dict == NULL) {
             dict = type->tp_dict = PyDict_New();
         }
-        END_TYPE_LOCK();
+        types_start_world();
         if (dict == NULL) {
             res = -1;
             goto done;
         }
     }
 
-    BEGIN_TYPE_DICT_LOCK(dict);
-    res = type_update_dict(type, (PyDictObject *)dict, name, value, &old_value);
-    assert(_PyType_CheckConsistency(type));
-    END_TYPE_DICT_LOCK();
+    if (is_dunder_name(name) && has_slotdef(name)) {
+        // The name corresponds to a type slot.
+        types_stop_world();
+        Py_BEGIN_CRITICAL_SECTION(dict);
```

Comment:
I don't think we should call `Py_BEGIN_CRITICAL_SECTION()` within a stop-the-world pause. I think it's ineffective at best and may possibly lead to deadlocks (not 100% sure).

Reply:
If that's true then we need to be extremely careful whatever APIs we call while the world is stopped.  Anything that uses a critical section internally could have the same issue, right?

The critical section on the dict is here because internally we call dict APIs that assert the critical section mutex is held.  We would need to create new APIs that don't assert that.  Seems messy to me.  I don't see why this shouldn't be okay.  When the world is stopped, all mutexes part of critical sections should be released.  So this should never block or do anything funny.   Or am I missing something?